Sources/Sentry/SentryThreadInspector.m

```diff
@@ -28,27 +62,89 @@ - (id)initWithStacktraceBuilder:(SentryStacktraceBuilder *)stacktraceBuilder
 {
     NSMutableArray<SentryThread *> *threads = [NSMutableArray new];
 
-    SentryCrashMC_NEW_CONTEXT(context);
-    [self.machineContextWrapper fillContextForCurrentThread:context];
+    @synchronized(self) {
+        SentryCrashMC_NEW_CONTEXT(context);
+        SentryCrashThread currentThread = sentrycrashthread_self();
+
+        [self.machineContextWrapper fillContextForCurrentThread:context];
+        int threadCount = [self.machineContextWrapper getThreadCount:context];
+
+        for (int i = 0; i < threadCount; i++) {
+            SentryCrashThread thread = [self.machineContextWrapper getThread:context withIndex:i];
+            SentryThread *sentryThread = [[SentryThread alloc] initWithThreadId:@(i)];
+
+            sentryThread.name = [self getThreadName:thread];
+
+            sentryThread.crashed = @NO;
+            bool isCurrent = thread == currentThread;
+            sentryThread.current = @(isCurrent);
+
+            if (isCurrent) {
+                sentryThread.stacktrace = [self.stacktraceBuilder buildStacktraceForCurrentThread];
+            }
+
+            // We need to make sure the main thread is always the first thread in the result
+            if ([self.machineContextWrapper isMainThread:thread])
+                [threads insertObject:sentryThread atIndex:0];
+            else
+                [threads addObject:sentryThread];
+        }
+    }
+
+    return threads;
+}
+
+- (NSArray<SentryThread *> *)getCurrentThreadsWithStackTrace
+{
+    NSMutableArray<SentryThread *> *threads = [NSMutableArray new];
 
-    int threadCount = [self.machineContextWrapper getThreadCount:context];
+    @synchronized(self) {
+        SentryCrashMC_NEW_CONTEXT(context);
+        SentryCrashThread currentThread = sentrycrashthread_self();
 
-    for (int i = 0; i < threadCount; i++) {
-        SentryCrashThread thread = [self.machineContextWrapper getThread:context withIndex:i];
-        SentryThread *sentryThread = [[SentryThread alloc] initWithThreadId:@(i)];
+        thread_act_array_t suspendedThreads = nil;
```

Comment:
do you have to explicitly free this, or does the obj-c runtime automatically do that for you?

Reply:
`sentrycrashmc_resumeEnvironment` called in Line 123 is responsible to free this array.